### PR TITLE
Refactored hook handlers in mongoose_metrics_mam_hooks

### PR DESCRIPTION
**Pull request by pawlooss1**
This PR changes all hook handlers in `mongoose_metrics_mam_hooks` module to `gen_hook` format.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/147438) / small_tests / 6125938c25ea296dd53a803de3f8aa9150f612d3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3829/147438/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3829/147438/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-27_09.11.18/lib.mongooseim.logs/run.2022-10-27_09.11.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/147489) / small_tests / dcd59295ab30d1a2f7ab2811e44441d802c5d3a7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3829/147489/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3829/147489/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-27_10.27.50/lib.mongooseim.logs/run.2022-10-27_10.27.50/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/147603) / small_tests / 3bb18b8fb98525a3982875ceb871b7cf3de2d6a4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3829/147603/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3829/147603/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-28_07.53.15/lib.mongooseim.logs/run.2022-10-28_07.53.16/suite.log.html)

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 1]
```diff
@@ -16,80 +16,77 @@
...
-                    Host :: jid:server(),
-                    _ArcID :: mod_mam:archive_id(),
-                    _ArcJID :: jid:jid()) -> any().
-mam_get_prefs(Result, Host, _ArcID, _ArcJID) ->
+-spec mam_get_prefs(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: any(),
+    Params :: map(),
+    Extra :: #{host_type := mongooseim:host_type()}.
+mam_get_prefs(Acc, _, #{host_type := Host}) ->
     mongoose_metrics:update(Host, modMamPrefsGets, 1),
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:65]: We might rename these variables to `HostType`, it is what they actually are but were just not updated when we introduced host types 🙂

**Review by NelsonVides — COMMENTED** [review]
Just one minor comment.

File: src/metrics/mongoose_metrics_mam_hooks.erl (hunk 1 above)
Comment (pawlooss1) [reply src/metrics/mongoose_metrics_mam_hooks.erl]: done 👍

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/148265) / small_tests / 41540df7e96d00d13fe30edf3ff9d96730a71750
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3829/148265/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3829/148265/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-02_09.19.03/lib.mongooseim.logs/run.2022-11-02_09.19.04/suite.log.html)

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 2]
```diff
@@ -16,135 +16,158 @@
...
-                         _ArcID :: mod_mam:archive_id(),
-                         _ArcJID :: jid:jid()) -> metrics_notify_return().
-mam_remove_archive(Acc, Host, _ArcID, _ArcJID) ->
-    mongoose_metrics:update(Host, modMamArchiveRemoved, 1),
-    Acc.
-
+-spec mam_get_prefs(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: any(),
+    Params :: map(),
+    Extra :: #{host_type := mongooseim:host_type()}.
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:63]: I think that all these declaring a map with the host_type are a bit repetitive. It should perhaps be an exported type from `gen_hook`. For example, see how I did it in the c2s feature branch here: https://github.com/esl/MongooseIM/blob/49eb66ceb5bb6df8a0d8883916c42a2524cdd232/src/gen_hook.erl#L35 Then hooks can use that exported type and reduce duplication. Feel free to copy-paste those, I'll deal with the merges 👌🏽

**Review by NelsonVides — COMMENTED** [review]
Just one comment, to improve duplication.

File: src/metrics/mongoose_metrics_mam_hooks.erl (hunk 2 above)
Comment (pawlooss1) [reply src/metrics/mongoose_metrics_mam_hooks.erl]: done 👍

**Review by NelsonVides — APPROVED** [review]
Excellent 👌🏽

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149359) / small_tests / 4b6dc147f8f8c9347d9df7fb94a7157245e96c37
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3829/149359/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3829/149359/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_11.28.19/lib.mongooseim.logs/run.2022-11-07_11.28.19/suite.log.html)

**Pull request merged**